google/cloud/spanner/internal/transaction_test.cc

```diff
@@ -0,0 +1,156 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/transaction.h"
+#include "google/cloud/spanner/internal/time.h"
+#include "google/cloud/spanner/timestamp.h"
+#include "google/cloud/spanner/transaction.h"
+#include <gmock/gmock.h>
+#include <chrono>
+#include <ctime>
+#include <future>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+using TransactionSelector = google::spanner::v1::TransactionSelector;
+
+class KeySet {};
+class ResultSet {};
+
+// Fake `google::cloud::spanner::Client` that supports a single `Read()`
+// operation that does nothing but track the expected transaction callbacks.
+class Client {
+ public:
+  // Set the `read_timestamp` we expect to see, and the `txn_id` we want to
+  // use during the upcoming `Read()` calls.
+  void Reset(Timestamp read_timestamp, std::string txn_id) {
+    read_timestamp_ = read_timestamp;
+    txn_id_ = std::move(txn_id);
+    std::unique_lock<std::mutex> lock(mu_);
+    valid_visits_ = 0;
+  }
+
+  // Return the number of valid visitations made to the transaction during a
+  // completed set of `Read()` calls.
+  int ValidVisits() {
+    std::unique_lock<std::mutex> lock(mu_);
+    return valid_visits_;
+  }
+
+  // User-visible read operation.
+  ResultSet Read(Transaction txn, std::string const& table, KeySet const& keys,
+                 std::vector<std::string> const& columns) {
+    std::function<ResultSet(TransactionSelector&)> read =
+        [this, &table, &keys, &columns](TransactionSelector& selector) {
+          return this->Read(selector, table, keys, columns);
+        };
+    return txn->Visit(std::move(read));
+  }
+
+ private:
+  ResultSet Read(TransactionSelector& selector, std::string const& table,
+                 KeySet const& keys, std::vector<std::string> const& columns);
+
+  Timestamp read_timestamp_;
+  std::string txn_id_;
+  std::mutex mu_;
+  int valid_visits_;  // GUARDED_BY(mu_);
+};
+
+// Transaction callback.  Normally we would use the TransactionSelector
+// to make a StreamingRead() RPC, and then, if the selector was a `begin`,
+// switch the selector to use the allocated transaction ID.  Here we use
+// the pre-assigned transaction ID after checking the read timestamp.
+ResultSet Client::Read(TransactionSelector& selector, std::string const&,
+                       KeySet const&, std::vector<std::string> const&) {
+  if (selector.has_begin()) {
```

Comment:
Should we collapse the nested `if()` block to a single boolean expression?

Reply:
Done.